src/ray/observability/tests/ray_event_recorder_test.cc

```diff
@@ -0,0 +1,111 @@
+    }
+    callback(Status::OK(), rpc::events::AddEventsReply{});
+  }
+
+ private:
+  std::vector<rpc::events::RayEvent> &recorded_events_;
+};
+
+class RayEventRecorderTest : public ::testing::Test {
+ public:
+  RayEventRecorderTest() {
+    mock_client_ = std::make_unique<MockEventAggregatorClient>(recorded_events_);
+    recorder_ = std::make_unique<RayEventRecorder>(*mock_client_, io_service_);
+    recorder_->StartExportingEvents();
+  }
+
+  instrumented_io_context io_service_;
+  std::vector<rpc::events::RayEvent> recorded_events_;
+  std::unique_ptr<MockEventAggregatorClient> mock_client_;
+  std::unique_ptr<RayEventRecorder> recorder_;
+};
+
+TEST_F(RayEventRecorderTest, TestRecordEvents) {
+  rpc::JobTableData data1;
+  data1.set_job_id("test_job_id_1");
+  data1.set_is_dead(false);
+  data1.set_driver_pid(12345);
+  data1.set_start_time(absl::ToUnixSeconds(absl::Now()));
+  data1.set_end_time(0);
+  data1.set_entrypoint("python test_script.py");
+  data1.mutable_driver_address()->set_ip_address("127.0.0.1");
+
+  rpc::JobTableData data2;
+  data2.set_job_id("test_job_id_2");
+  data2.set_is_dead(true);
+  data2.set_driver_pid(67890);
+  data2.set_start_time(absl::ToUnixSeconds(absl::Now()) - 3600);  // 1 hour ago
+  data2.set_end_time(absl::ToUnixSeconds(absl::Now()));
+  data2.set_entrypoint("python another_script.py");
+  data2.mutable_driver_address()->set_ip_address("192.168.1.100");
+
+  std::vector<std::unique_ptr<RayEventInterface>> events;
+  events.push_back(
+      std::make_unique<RayDriverJobDefinitionEvent>(data1, "test_session_name_1"));
+  events.push_back(std::make_unique<RayDriverJobExecutionEvent>(
+      data2, rpc::events::DriverJobExecutionEvent::SUCCESS, "test_session_name_2"));
+  recorder_->AddEvents(std::move(events));
+  io_service_.run_for(
+      std::chrono::milliseconds(RayConfig::instance().ray_events_report_interval_ms()));
+  ASSERT_EQ(recorded_events_.size(), 2);
+
+  // Verify first event
+  ASSERT_EQ(recorded_events_[0].source_type(), rpc::events::RayEvent::GCS);
+  ASSERT_EQ(recorded_events_[0].session_name(), "test_session_name_1");
+  ASSERT_EQ(recorded_events_[0].event_type(),
+            rpc::events::RayEvent::DRIVER_JOB_DEFINITION_EVENT);
+  ASSERT_EQ(recorded_events_[0].severity(), rpc::events::RayEvent::INFO);
+  ASSERT_TRUE(recorded_events_[0].has_driver_job_definition_event());
+  ASSERT_EQ(recorded_events_[0].driver_job_definition_event().job_id(), "test_job_id_1");
+
+  // Verify second event
+  ASSERT_EQ(recorded_events_[1].source_type(), rpc::events::RayEvent::GCS);
+  ASSERT_EQ(recorded_events_[1].session_name(), "test_session_name_2");
+  ASSERT_EQ(recorded_events_[1].event_type(),
+            rpc::events::RayEvent::DRIVER_JOB_EXECUTION_EVENT);
+  ASSERT_EQ(recorded_events_[1].severity(), rpc::events::RayEvent::INFO);
+  ASSERT_TRUE(recorded_events_[1].has_driver_job_execution_event());
+  ASSERT_EQ(recorded_events_[1].driver_job_execution_event().job_id(), "test_job_id_2");
+}
+
```

Comment:
Do you have tests for the merge logic

Reply:
We don't need to test at the ray_event_recorder level. We should just unit test RayEvent